cmd/snap-recovery-chooser/main.go

```diff
@@ -0,0 +1,198 @@
+// {
+//     "label": "<system-label",
+//     "action": {} // client.SystemAction object
+// }
+//
+// No action is forwarded to snapd if the chooser UI exits with an error code or
+// the response structure is invalid.
+//
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+)
+
+var (
+	// default marker file location
+	defaultMarkerFile = "/run/snapd-recovery-chooser-triggered"
+
+	Stdout io.Writer = os.Stdout
+	Stderr io.Writer = os.Stderr
+
+	chooserTool = consoleConfWrapperUITool
+)
+
+// consoleConfWrapperUITool returns a hardcoded path to the console conf wrapper
+func consoleConfWrapperUITool() (*exec.Cmd, error) {
+	tool := filepath.Join(dirs.GlobalRootDir, "usr/bin/console-conf")
+
+	if _, err := os.Stat(tool); err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("chooser UI tool %q does not exist", tool)
+		}
+		return nil, fmt.Errorf("cannot stat UI tool binary: %v", err)
+	}
+	// TODO:UC20 update once upstream console-conf has the right command
+	// line switches
+	return exec.Command(tool, "--recovery-chooser-mode"), nil
+}
+
+// ChooserSystems carries the list of available recovery systems
+type ChooserSystems struct {
+	Systems []client.System `json:"systems,omitempty"`
+}
+
+func outputForUI(out io.Writer, sys *ChooserSystems) error {
+	enc := json.NewEncoder(out)
+	if err := enc.Encode(sys); err != nil {
+		return fmt.Errorf("cannot serialize chooser options: %v", err)
+	}
+	return nil
+}
+
+// Response is sent by the UI tool and contains the choice made by the user
+type Response struct {
+	Label  string              `json:"label"`
+	Action client.SystemAction `json:"action"`
+}
+
+func runUI(cmd *exec.Cmd, sys *ChooserSystems) (rsp *Response, err error) {
+	var asBytes bytes.Buffer
+	if err := outputForUI(&asBytes, sys); err != nil {
+		return nil, err
+	}
+
+	logger.Noticef("spawning UI")
+	// the UI uses the same tty as current process
+	cmd.Stdin = &asBytes
+	// reuse stderr
+	cmd.Stderr = os.Stderr
+	// the chooser may be invoked via console-conf service which uses
+	// KillMethod=process, so make sure the UI process dies when we die
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Pdeathsig: syscall.SIGKILL,
+	}
+
+	out, err := cmd.Output()
+	if err != nil {
+		return nil, fmt.Errorf("cannot collect output of the UI process: %v", err)
+	}
+
+	logger.Noticef("UI completed")
+
+	var resp Response
+	dec := json.NewDecoder(bytes.NewBuffer(out))
+	if err := dec.Decode(&resp); err != nil {
+		return nil, fmt.Errorf("cannot decode response: %v", err)
+	}
+	return &resp, nil
+}
+
+func cleanupTriggerMarker() error {
+	err := os.Remove(defaultMarkerFile)
+	if os.IsNotExist(err) {
+		return nil
+	}
+	return err
+}
+
+func chooser(cli *client.Client) error {
+	snappyTesting := os.Getenv("SNAPPY_TESTING_USE_STDOUT") != ""
```

Comment:
We have those "getenv bool" helpers. This looks like a place to use one.

Reply:
the SNAPPY_ prefix is historical, we use SNAPD_ these days, please let's not use it for new things

Reply:
I'll go with SNAPD_TESTING_USE_STDOUT.

Reply:
We could do something like https://paste.ubuntu.com/p/sXqWJkT7wp/ and check the outputForUI this way (or keeping in a separate test with something like https://paste.ubuntu.com/p/s86HQY8zfn/)